migrator/migrator.go

```diff
@@ -0,0 +1,272 @@
+		err = m.addNodeToEtcdCluster(ctx, nodeNames, 3)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+	} else if memberCount == 1 {
+		//  ensure that first node has proper etcd peer url set to etcd1.xxxx.xxxx.xxx
+		err = m.fixFirstNodePeerUrl(ctx, memberListResponse.Members)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		// add second node to the etcd cluster
+		err = m.addNodeToEtcdCluster(ctx, nodeNames, 2)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		// add third node to the etcd cluster
+		err = m.addNodeToEtcdCluster(ctx, nodeNames, 3)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	} else {
+		fmt.Printf("unexpected number of nodes in etcd cluster\n")
+		return microerror.Maskf(executionFailedError, fmt.Sprintf("found %d nodes in etcd cluster", memberCount))
+	}
+
+	fmt.Printf("ETCD cluster migration succesfuly finished.\n\n")
+	return nil
+}
+
+// fixFirstNodePeerUrl ensure the peerURL for the first node in etcdcluster is properly set
+// as it can have 'localhost' value from the previous version fo k8scloudconfig.
+func (m *Migrator) fixFirstNodePeerUrl(ctx context.Context, etcdMembers []*etcdserver.Member) error {
+	id := etcdMembers[0].ID
+	peerUrls := []string{etcdPeerName(m.etcdStartingIndex, m.baseDomain)}
+
+	_, err := m.etcdClient.Cluster.MemberUpdate(ctx, id, peerUrls)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	fmt.Printf("Updated first node PeerUrls to %s.\n", peerUrls)
+	return nil
+}
+
+// addNodeToEtcdCluster configure etcd3 service on the second or third node in order
+// to join the existing cluster via k8s job executed on the node and after that
+// it will add the node to the etcd cluster via etcdv3 client API.
+
+func (m *Migrator) addNodeToEtcdCluster(ctx context.Context, nodeNames []string, nodeCount int) error {
+	// nodeCount can only be 2 or 3
+	// 2 when adding second node to a single node etcd cluster
+	// 3 when adding third node to two node etcd cluster
+	if nodeCount != 2 && nodeCount != 3 {
+		return microerror.Maskf(executionFailedError, "nodeCount can only have values 2 or 3")
+	}
+
+	if len(nodeNames) != 3 {
+		return microerror.Maskf(executionFailedError, "nodeNames len must be 3")
+	}
+
+	// execute commands on the node to configure new etcd3 member so that it can join the existing cluster
+	{
+		nodeName := nodeNames[nodeCount-1]
+
+		// the final sed command may look like this:
+		// sed -i 's/--initial-cluster .*\\/--initial-cluster etcd1=https://etcd1.clusterd.domain.io:2380,etcd1=https://etcd2.clusterd.domain.io:2380/g' /etc/systemd/system/etcd3.service'
+		sedReplaceRegEx := "--initial-cluster .*\\\\"
+		sedReplaceWith := fmt.Sprintf("--initial-cluster %s\\\\", initialCluster(m.etcdStartingIndex, m.baseDomain, nodeCount))
+		sedInitialClusterCmd := fmt.Sprintf("sed -i 's/%s/%s/g' /etc/systemd/system/etcd3.service", sedReplaceRegEx, sedReplaceWith)
+
+		commands := []string{
```

Comment:
If you want, you can avoid the rimraf by creating another unit that is Required by the etcd3 unit and simply waits for an environment file to be present.
This way you don't have to edit the unit file but it works out of the box.

https://github.com/giantswarm/azure-operator/blob/ha-masters-vmss/service/controller/cloudconfig/temp.go#L182
https://github.com/giantswarm/azure-operator/blob/ha-masters-vmss/service/controller/cloudconfig/temp.go#L163

And in `/opt/bin/setup-etcd-environment` something like:

```
#!/bin/bash

while [ ! -f /etc/my-env-file ]
do
   echo "File not found"
   sleep 5
done
```

Then you simply create the file with the system you're using here.